### PR TITLE
14 feature fe upload code

### PR DESCRIPTION
**Pull request by ashan36**
## Done
- Basic code-upload page layout
- Implemented custom rich text editor
- Toggle buttons for toolbar
- Code block, blockquote, unordered list formatted block text
- Code block key commands
- Code block syntax highlighting based on language selector
- Submit button that pulls texteditor content out as JSON to send to the server

## To do
- Turns out the best way to bundle a configured prismjs file is to use babel-loader. In order to configure our build options, we'd need to run the eject script that comes from create-react-app. -- - - Alternatively we could just run it not as a dependency and package the entire thing up in our build. The whole thing is 44kb with the languages currently on our list.
- Set language drop down options based on user experience object.

### CODE REVIEW

**Comment by ashan36** [comment]
If you are testing the syntax highlighting, only the grammar for Javascript is loaded by default in the dependency. If you want the others you will have to download and replace prismjs with the one from the link here: https://prismjs.com/download.html#themes=prism-okaidia&languages=markup+css+clike+javascript+c+csharp+cpp+java+python+ruby

File: client/src/pages/CodeUpload.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+} from "@material-ui/core";
+import { UserContext } from "context/UserContext";
+import { TextEditor } from "components/index";
+import ErrorMessage from "components/ErrorMessage";
+import SubmitButton from "components/TextEditor/components/SubmitButton";
+
+const useStlyes = makeStyles({
+  root: {
+    padding: "5%"
+  },
```
Comment (rkerins91) [inline client/src/pages/CodeUpload.js:23]: I think the background here looks out of place from the rest of out site. Can add `backgroundImage: "linear-gradient(to top, rgba(87,40,184,1) 25%, rgba(119,140,255,1) 100%)",` to root to get the gradient and make the site look uniform. Also probably want to have the corners rounded like in Nick's SignUpContainer component.

**Review by rkerins91 — DISMISSED** [review]

File: client/src/pages/CodeUpload.js (hunk 1 above)
Comment (ashan36) [reply client/src/pages/CodeUpload.js:23]: Rounded the corners.

I chose that background color because I assumed from the RFP that the main site color would be this (see profile, adding credits, and reviews mockups), and only the onboarding, login, and signup pages have the purple gradient. We can discuss more on the site styling in standup.

File: client/src/App.css [hunk 2]
```diff
@@ -31,3 +31,17 @@
...
   }
 }
+
+/* .public-DraftStyleDefault-pre {
+  background-color: rgba(0, 0, 0, 0.05);
+  font-family: "Inconsolata", "Menlo", "Consolas", monospace;
+  font-size: 16px;
+  padding: 20px;
+  margin-bottom: 20px;
+} */
```
Comment (skassam21) [inline client/src/App.css:41]: Remove this commented out code

File: client/src/App.js [hunk 3]
```diff
@@ -2,11 +2,13 @@ import React, { useState, useMemo } from "react";
 import { MuiThemeProvider } from "@material-ui/core";
 import { BrowserRouter, Route } from "react-router-dom";
 import { UserContext } from "./context/UserContext";
+import { SnackbarProvider } from "notistack";
```
Comment (skassam21) [inline client/src/App.js:5]: Why are we using a snackbar library when material-ui has snackbars already - this will add unnecessary bloat to the application

File: client/src/components/TextEditor/TextEditor.js [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  convertToRaw
+} from "draft-js";
+
+import CodeUtils from "draft-js-code";
+import Prism from "prismjs";
+import PrismDecorator from "draft-js-prism";
+import "prismjs/themes/prism-okaidia.css";
+
+const TextEditor = props => {
+  const useStlyes = makeStyles({
```
Comment (skassam21) [inline client/src/components/TextEditor/TextEditor.js:19]: misspelled useStyles

File: client/src/components/TextEditor/TextEditor.js [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+  RichUtils,
+  convertToRaw
+} from "draft-js";
+
+import CodeUtils from "draft-js-code";
+import Prism from "prismjs";
+import PrismDecorator from "draft-js-prism";
+import "prismjs/themes/prism-okaidia.css";
+
+const TextEditor = props => {
```
Comment (skassam21) [inline client/src/components/TextEditor/TextEditor.js:18]: I generally like to destructure props, but that's a preference, some people don't like to do that

File: client/src/components/TextEditor/TextEditor.js [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    focusEditor();
+  }, []);
+
+  //Toggle button group controller
+  const handleFormatChange = style => {
+    if (
+      style === "BOLD" ||
+      style === "ITALIC" ||
+      style === "UNDERLINE" ||
+      style === "CODE"
```
Comment (skassam21) [inline client/src/components/TextEditor/TextEditor.js:104]: These strings should be constants, possibly at the top of your file (outside the component)

File: client/src/components/TextEditor/TextEditor.js [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+  const editorStyleMap = {
+    CODE: {
+      background: "#414239",
+      fontFamily:
+        "'Consolas', 'Monaco', 'Andale Mono', 'Ubuntu Mono', monospace",
+      fontSize: "1em",
+      textShadow: "0 1px rgba(0, 0, 0, 0.3)",
+      color: "#f8f8f2"
+    }
+  };
```
Comment (skassam21) [inline client/src/components/TextEditor/TextEditor.js:42]: This probably doens't need to be defined in the component itself

File: client/src/components/TextEditor/TextEditor.js [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+      setLanguageState(props.selectedLanguage);
+    } else {
+      //Null is a fallback grammar for Prism
+      setLanguageState(null);
+    }
+  }, [props.selectedLanguage]);
+
+  const decorator = new PrismDecorator({
+    prism: Prism,
+    defaultSyntax: languageState
```
Comment (skassam21) [inline client/src/components/TextEditor/TextEditor.js:57]: why is languageState a state variable... it functions the same as props.selectedLanguage?

Comment (skassam21) [reply client/src/components/TextEditor/TextEditor.js:57]: Also, I wonder if this needs to be keep redefined (I don't know how costly it is to keep redefining it), every time the component renders

File: client/src/components/TextEditor/components/SubmitButton.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import React, { useState } from "react";
+import { Button, makeStyles } from "@material-ui/core";
+
+const SubmitButton = props => {
+  const useStyles = makeStyles({
+    button: {
+      backgroundColor: "#43DDC1",
```
Comment (skassam21) [inline client/src/components/TextEditor/components/SubmitButton.js:7]: Why do we have this specific custom SubmitButton - background colours for buttons generally should be coming from the theme

File: client/src/components/TextEditor/components/Toolbar.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  const [inlineFormatButtonState, setInlineFormatButtonState] = useState();
+  const [blockFormatButtonState, setBlockFormatButtonState] = useState();
+
+  useEffect(() => {
+    setInlineFormatButtonState(props.InlineStyle);
+  }, [props.InlineStyle]);
+
+  useEffect(() => {
+    setBlockFormatButtonState(props.BlockStyle);
+  }, [props.BlockStyle]);
```
Comment (skassam21) [inline client/src/components/TextEditor/components/Toolbar.js:23]: Again, I think this is overusing state - you can just use the props values

File: client/src/pages/CodeUpload.js [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+                </MenuItem>
+              );
+            })}
+          </Select>
+        </Grid>
+        <Grid item xs={12}>
+          <TextEditor
+            selectedLanguage={requestLanguage}
+            onSubmit={handleSubmit}
+            didSubmit={submitState}
```
Comment (skassam21) [inline client/src/pages/CodeUpload.js:179]: The flow of submitting is confusing and hard to follow, it could probably be a bit better designed

**Review by skassam21 — COMMENTED** [review]
I know this is merged already, but I thought I'd give my feedback for things you can think about going forward

File: client/src/App.js (hunk 3 above)
Comment (ashan36) [reply client/src/App.js:5]: notistack is a wrapper for the material-ui snackbar component that allows for nicely displayed stackable snackbars. It saves a lot of customizing work to get multiple snackbars playing nice with eachother ourselves.
